coderd/workspaceapps/token.go

```diff
@@ -0,0 +1,209 @@
+package workspaceapps
+
+import (
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"time"
+
+	"github.com/go-jose/go-jose/v3"
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/coderd/database"
+)
+
+const (
+	tokenSigningAlgorithm     = jose.HS512
+	apiKeyEncryptionAlgorithm = jose.A256GCMKW
+)
+
+// SignedToken is the struct data contained inside a workspace app JWE. It
+// contains the details of the workspace app that the token is valid for to
+// avoid database queries.
+type SignedToken struct {
+	// Request details.
+	Request `json:"request"`
+
+	// Trusted resolved details.
+	Expiry      time.Time `json:"expiry"` // set by GenerateToken if unset
+	UserID      uuid.UUID `json:"user_id"`
+	WorkspaceID uuid.UUID `json:"workspace_id"`
+	AgentID     uuid.UUID `json:"agent_id"`
+	AppURL      string    `json:"app_url"`
+}
+
+// MatchesRequest returns true if the token matches the request. Any token that
+// does not match the request should be considered invalid.
+func (t SignedToken) MatchesRequest(req Request) bool {
+	return t.AccessMethod == req.AccessMethod &&
+		t.BasePath == req.BasePath &&
+		t.UsernameOrID == req.UsernameOrID &&
+		t.WorkspaceNameOrID == req.WorkspaceNameOrID &&
+		t.AgentNameOrID == req.AgentNameOrID &&
+		t.AppSlugOrPort == req.AppSlugOrPort
+}
+
+// AppSigningKey is used for signing and encrypting app tokens and API keys.
+//
+// The first 64 bytes of the key are used for signing tokens with HMAC-SHA256,
+// and the last 32 bytes are used for encrypting API keys with AES-256-GCM.
+type SigningKey [96]byte
```

Comment:
We should rename this now that it does both signing/encryption.

How about, "SecurityKey"?

Reply:
Rename is here: https://github.com/coder/coder/pull/6998/commits/fe28e42220513eb5007fca5b861d2427a0c5845f